File: commands/shards.js
import { EmbedBuilder, inlineCode } from 'discord.js';
import colors from '../models/colors';
import shardNames from '../models/shardNames';

const sum = values => values.reduce((p, v) => p + v, 0);

export const helpers = {
	ephemeral: true,
};

export async function run(interaction) {
	const client = interaction.client;
	const embed = new EmbedBuilder()
		.setAuthor({
			name: client.user.username,
			iconURL: client.user.avatarURL(),
			url: 'https://server-discord.com',
		})
		.setFooter({ text: `Шард сервера: ${shardNames[client.shard.ids[0]]}` })
		.setColor(colors.blue);

	const [guilds, pings, memory] = await Promise.all([
		client.shard.fetchClientValues('guilds.cache.size'),
		client.shard.broadcastEval(c => c.ws.ping.toFixed(0)),
		client.shard.broadcastEval(() => +(process.memoryUsage().heapUsed / 1024 / 1024).toFixed(2)),
	]);

	let description = '';
	for (let i = 0, length = client.shard.count; i < length; i++) {
		description +=
			`${i + 1}. ${shardNames[i]} ${i === client.shard.ids[0] ? '←' : ''}\n` +
			`Серверов: ${inlineCode(guilds[i])}, Пинг: ${inlineCode(pings[i])} мс, ОЗУ: ${inlineCode(memory[i])} МБ\n`;
	}

	embed.setDescription(description);
	embed.addFields([
		{ name: 'Всего', value: `Серверов: ${inlineCode(sum(guilds))}, ОЗУ: ${inlineCode(sum(memory).toFixed(2))} МБ` },
	]);
	await interaction.editReply({ embeds: [embed], ephemeral: interaction.options.getBoolean('ephemeral') ?? true });
}

export default {
	helpers,
	run,
};

File: commands/shards.go
package commands

import (
	"fmt"

	"github.com/bwmarrin/discordgo"

	"sdc/models"
)

func Shards(d *Deps, s *discordgo.Session, i *discordgo.InteractionCreate) error {
	snaps := d.Runtime.ShardSnapshots()
	botUser := s.State.User
	author := &discordgo.MessageEmbedAuthor{Name: "SDC", URL: "https://server-discord.com"}
	if botUser != nil {
		author.Name = botUser.Username
		author.IconURL = botUser.AvatarURL("")
	}

	desc := ""
	totalGuilds := 0
	totalMem := 0.0
	for _, sn := range snaps {
		mark := ""
		if sn.ID == s.ShardID {
			mark = " ←"
		}
		desc += fmt.Sprintf("%d. %s%s\nСерверов: `%d`, Пинг: `%d` мс, ОЗУ: `%.2f` МБ\n",
			sn.ID+1, models.ShardName(sn.ID), mark, sn.Guilds, sn.PingMS, sn.HeapMB)
		totalGuilds += sn.Guilds
		totalMem += sn.HeapMB
	}

	embed := &discordgo.MessageEmbed{
		Author:      author,
		Color:       models.EmbedColors["blue"],
		Description: desc,
		Footer: &discordgo.MessageEmbedFooter{
			Text: "Шард сервера: " + models.ShardName(s.ShardID),
		},
		Fields: []*discordgo.MessageEmbedField{{
			Name:  "Всего",
			Value: fmt.Sprintf("Серверов: `%d`, ОЗУ: `%.2f` МБ", totalGuilds, totalMem),
		}},
	}

	ephemeral, set := optionBool(i, "ephemeral")
	if !set {
		ephemeral = true
	}
	_ = ephemeral // already deferred with correct flag

	return editReply(s, i, &discordgo.WebhookEdit{Embeds: &[]*discordgo.MessageEmbed{embed}})
}
